### PR TITLE
Move project configuration to pyproject.toml

### PR DESCRIPTION
**Pull request by singingwolfboy**
The `pyproject.toml` file is the new standard for project configuration in Python. [It is the preferred configuration file for `pip`](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/), and many other tools support it as well. This pull request moves as much of the project configuration as possible into this file, deleting the now-unnecessary files that they used to live in.

Note that this pull request also deletes `setup.cfg` and `setup.py`! For more information about how this project can be packaged and uploaded to PyPI, [see the `flit` documentation](https://flit.pypa.io/en/stable/index.html). `flit` handles the heavy lifting for interfacing with PyPI. (Users do _not_ need `flit` installed on their computer in order to install this project! As I said, `pip` supports this build system, and has for many years now.)

### CODE REVIEW

**Comment by singingwolfboy** [comment]
Hi @MartinThoma, could you please review/merge this pull request?

**Comment by MartinThoma** [comment]
I'm hesitating to merge this one. The current workflow works and I don't see any issues with that.

I'm open to move isort ([docs](https://pycqa.github.io/isort/docs/configuration/config_files.html#pyprojecttoml-preferred-format)) and pytest ([docs](https://docs.pytest.org/en/6.2.x/customize.html#pyproject-toml)).

I'm way more hesitant with anything that could break packaging.

**Comment by MartinThoma** [comment]
Moving the coverage config is also not a big deal for me

**Comment by singingwolfboy** [comment]
I understand your hesitation, but I'm trying to move the Python developer community towards the new standard. I'll mention again that [`pyproject.toml` is the preferred packaging format for `pip`](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/), and `setup.py` is now considered "legacy".

Of course, this is your project, and it's your decision. But is there anything I can do to relieve your fears and help you migrate to the new standard?

**Comment by MasterOdin** [comment]
> I'm way more hesitant with anything that could break packaging.

Perhaps it'd help to hear your workflow for doing packaging and releasing? I think you use [`build`](https://pypi.org/project/build/) in combination with [`twine`](https://pypi.org/project/twine/) to handle publishing?

By listing out the explicit steps, can then show what that'd look like via flit (or whatever) instead with the `pyproject.toml` file.

**Comment by MartinThoma** [comment]
When I release a new version, I do this:

```
python make_changelog.py
# manually adjust the changelog
# manually bump the version in _version.py
make upload
git add CHANGELOG.md
git add PyPDF2/_version.py
git commit
# Add the changelog as a commit message with the subject line 'REL: x.y.z'
git push
git tag -s x.y.z
# Add the changelog as a tag message
git push --tags
# Create a release version on Github
```

The only thing that would change is the command behind `make upload`. How would that change?

> setup.py is now considered "legacy" [by pip]

Huh, that is interesting. I didn't know that. Thanks for sharing!

**Comment by MartinThoma** [comment]
@MasterOdin If I understand your comment right, you would also be in favor of merging this PR?

**Comment by singingwolfboy** [comment]
> The only thing that would change is the command behind `make upload`. How would that change?

Currently, `make upload` runs `python setup.py sdist bdist_wheel && twine upload -s dist/*`. That would change to just [`flit publish`](https://flit.pypa.io/en/stable/cmdline.html#flit-publish), which takes care of both building and uploading the artifacts to PyPI.

I see that you have `twine` listed as a dev dependency via `requirements/dev.txt`. Let me make some changes to this pull request so that you can install the dev dependencies by running `pip install -e .[dev]`, as well. I'll also rebase to resolve the conflicts.

**Comment by MasterOdin** [comment]
My comment was more that I think on merging this, it would ideally be that your workflow would ideally not change, or become simpler, and that @singingwolfboy PR would include this. So from the above, the question would be if `make upload` would continue to work as-is, or if that'd need to be changed. Perhaps other bits of the makefile (or gh action files) need changing as well, especially with the mention of [flit](https://pypi.org/project/flit/).

As for whether this should be done, I would classify myself as largely "neutral" in merging this. I think that the goal of consolidating around `pyproject.toml` in the ecosystem will be good and at some point you'll probably have to, but also that the rollout of it across the python ecosystem has been a huge dumpster fire. The adoption of this format has been over a half-decade in the process at this point, and some aspects of it still aren't really standardized in a satisfying way, and then you've got poetry which doesn't follow [PEP-621](https://peps.python.org/pep-0621/) and probably won't for a while yet.

Then you've got stuff like the fact that pip recommends everyone uses pyproject.toml, but still uses the legacy setup.py setup (amongst other very popular projects) and so doesn't necessarily inspire confidence that the format won't change in the future requiring maintenance overhead...

**Comment by MartinThoma** [comment]
> Currently, make upload runs python setup.py sdist bdist_wheel && twine upload -s dist/*. That would change to just [flit publish](https://flit.pypa.io/en/stable/cmdline.html#flit-publish)

How can I inspect the artifact flit creates before / without uploading it?

**Comment by singingwolfboy** [comment]
> How can I inspect the artifact flit creates before / without uploading it?

You can use [`flit build`](https://flit.pypa.io/en/stable/cmdline.html#flit-build) to build the artifact, and inspect it at that point. Once you're satisfied, you can run `flit publish` and it will upload that same artifact -- or if you want to be ~~paranoid~~ extra sure, there's nothing wrong with continuing to use `twine` to upload the built artifact, just as before. I assumed that you would want fewer tools in your workflow, so I removed `twine` from the `requirements/dev.in`, but I'm happy to revert that change if my assumption was wrong!

**Comment by singingwolfboy** [comment]
> My comment was more that I think on merging this, it would ideally be that your workflow would ideally not change, or become simpler, and that @singingwolfboy PR would include this. So from the above, the question would be if `make upload` would continue to work as-is, or if that'd need to be changed. Perhaps other bits of the makefile (or gh action files) need changing as well, especially with the mention of [flit](https://pypi.org/project/flit/).

Indeed! Thank you for pointing that out -- I believe I've made the necessary changes in this PR so that the workflow should be unchanged.
 
> As for whether this should be done, I would classify myself as largely "neutral" in merging this. I think that the goal of consolidating around `pyproject.toml` in the ecosystem will be good and at some point you'll probably have to, but also that the rollout of it across the python ecosystem has been a huge dumpster fire. The adoption of this format has been over a half-decade in the process at this point, and some aspects of it still aren't really standardized in a satisfying way, and then you've got poetry which doesn't follow [PEP-621](https://peps.python.org/pep-0621/) and probably won't for a while yet.
> 
> Then you've got stuff like the fact that pip recommends everyone uses pyproject.toml, but still uses the legacy setup.py setup (amongst other very popular projects) and so doesn't necessarily inspire confidence that the format won't change in the future requiring maintenance overhead...

True. To be fair, packaging in the Python ecosystem has been more-or-less a dumpster fire for roughly as long as the Python ecosystem has existed. To me, `pyproject.toml` feels like a step in the right direction.

**Comment by MasterOdin** [comment]
> True. To be fair, packaging in the Python ecosystem has been more-or-less a dumpster fire for roughly as long as the Python ecosystem has existed. To me, pyproject.toml feels like a step in the right direction.

Agreed on both points.

**Comment by MartinThoma** [comment]
I'm currently checking the distribution files. Left is old, right is new.

## Wheel

I changed the order of entries on the left so that the diff is easier to read

image

* Author and maintainer name/email is different: OK
* License is no longer included :warning: Can we fix that? Did the Python community find another way to standardize how to communicate licenses?
* `Platform: UNKNOWN` was removed: OK. I have no idea why it was there in the first place
* `Requires-Dist`: Spaces and quotes changed: OK
* `Requires-Dist`: `dev` and `docs` was added. Nice!
* `Home-page:` changed to a project-url: OK

image

The file `top_level.txt` which contains just `PyPDF2` is no longer there. Seems to be ok: https://github.com/python-poetry/poetry/issues/3093#issuecomment-772411398

image

**Comment by MartinThoma** [comment]
:warning: Blocker for merging: The `sdist` distribution file is empty

```
$ python -m venv venv
$ source venv/bin/activate
$ pip install flit
$ flit build --format sdist
/home/moose/.pyenv/versions/3.10.2/lib/python3.10/site-packages/requests/__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.12) or chardet (5.0.0)/charset_normalizer (2.0.11) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({})/charset_normalizer ({}) doesn't match a supported "
Untracked or deleted files in the source directory. Commit, undo or ignore these files in your VCS.
$ ls -lh dist
total 4,0K
-rw-rw-r-- 1 moose moose 63 Okt 22 10:23 PyPDF2-2.11.1.tar.gz
```

**Comment by MartinThoma** [comment]
## sdist: Needs adjustment

image

The new sdist should look more similar to the old one. That means the following files/folders should not be added:
 `.github`, `docs`, `requirements`, `resources`, `sample-files`, `tests`, `.gitmodules`, `.gitignore`, `.git-blame-ignore-revs`, `.pre-commit-config.yml`, `.flake8`, `.pylintrc`, `Makefile`, `make_changelog.py`, `mutmut-test.sh`, `mutmut_config.py`, `tox.ini`

**Comment by MartinThoma** [comment]
For the license, I see https://flit.pypa.io/en/stable/flit_ini.html?highlight=license#metadata-section

> License: The name of a license, if you’re using one for which there isn’t a Trove classifier. It’s recommended to use Trove classifiers instead of this in most cases.

[PEP 639](https://peps.python.org/pep-0639/) (draft status) seems very relevant. I need to read that.

**Comment by MartinThoma** [comment]
@singingwolfboy Thank you for all the work you already put into this :pray: If you want to focus on other topics, I would understand that. In this case I could very likely figure out the rest by myself (in a follow-up change). Just let me know how you would like to proceed :-)

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "PyPDF2"
+authors = [{ name = "Mathieu Fenniak", email = "biziqe@mathieu.fenniak.net" }]
+maintainers = [{ name = "Martin Thoma", email = "info@martin-thoma.de" }]
+description = "A pure-python PDF library capable of splitting, merging, cropping, and transforming PDF files"
+readme = "README.md"
+dynamic = ["version"]
+license = { file = "LICENSE" }
```
Comment (MartinThoma) [inline pyproject.toml:12]:
```suggestion
license = BSD-3-Clause
```

See https://python-poetry.org/docs/pyproject/#license

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+docs = ["sphinx", "sphinx_rtd_theme", "myst_parser"]
+
+[tool.mutmut]
+backup = false
+runner = "./mutmut-test.sh"
+tests_dir = "tests/"
+
+[tool.check-wheel-contents]
+package = "./PyPDF2"
+
```
Comment (MartinThoma) [inline pyproject.toml:53]:
```suggestion

[tool.flit.sdist]
exclude = [".github/*", "docs/*", "resources/*", "sample-files/*", "sample-files/.github/*", "sample-files/.gitignore", "sample-files/.pre-commit-config.yaml",  "requirements/*", "tests/*", ".flake8", ".gitignore", ".gitmodules", ".pylintrc", "tox.ini", "make_changelog.py", "mutmut-test.sh", ".pre-commit-config.yaml", ".gitblame-ignore-revs", "Makefile", "mutmut_config.py"]

```

**Comment by MartinThoma** [comment]
@singingwolfboy If you add the two suggested changes, the PR would be fine and could be merged as-is :-)

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "PyPDF2"
+authors = [{ name = "Mathieu Fenniak", email = "biziqe@mathieu.fenniak.net" }]
+maintainers = [{ name = "Martin Thoma", email = "info@martin-thoma.de" }]
+description = "A pure-python PDF library capable of splitting, merging, cropping, and transforming PDF files"
+readme = "README.md"
+dynamic = ["version"]
+license = BSD-3-Clause
```
Comment (MartinThoma) [inline pyproject.toml:12]:
```suggestion
license = "BSD-3-Clause"
```

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "PyPDF2"
+authors = [{ name = "Mathieu Fenniak", email = "biziqe@mathieu.fenniak.net" }]
+maintainers = [{ name = "Martin Thoma", email = "info@martin-thoma.de" }]
+description = "A pure-python PDF library capable of splitting, merging, cropping, and transforming PDF files"
+readme = "README.md"
+dynamic = ["version"]
+license = "BSD-3-Clause"
```
Comment (MartinThoma) [inline pyproject.toml:12]:
```suggestion
license = { file = "LICENSE" }
```

I'm very sorry about this back-and-forth. I know looked it up:

* https://flit.pypa.io/en/latest/pyproject_toml.html#metadata-section
* https://peps.python.org/pep-0621/#license
* https://python-poetry.org/docs/pyproject/#license

So you can see that poetry and flit define different types of values for this field. PEP 621 is not clear how it should be structured.

I've opened a question about this on PyPI discussions https://github.com/pypi/warehouse/discussions/12407

As flit does not accept a string-entry in here, we have to change it back to what you originally had.

**Comment by MartinThoma** [comment]
It's merged! Thank you @singingwolfboy for your work and for helping the Python community to move forward :heart:

**Comment by singingwolfboy** [comment]
> It's merged! Thank you @singingwolfboy for your work and for helping the Python community to move forward ❤️

You're very welcome! And thank _you_ for being so thorough with your review, and for pushing this PR the last few steps over the finish line. 😄

**Pull request merged**